packages/wrangler/src/__tests__/pages-deployment-tail.test.ts

```diff
@@ -0,0 +1,772 @@
+			setIsTTY(true);
+			const api = mockTailAPIs();
+			await runWrangler(
+				"pages deployment tail mock-deployment-id --project-name mock-project"
+			);
+
+			const event = generateMockRequestEvent();
+			const message = generateMockEventMessage({ event });
+			const serializedMessage = serialize(message);
+
+			api.ws.send(serializedMessage);
+			expect(
+				std.out
+					.replace(
+						new Date(mockEventTimestamp).toLocaleString(),
+						"[mock event timestamp]"
+					)
+					.replace(
+						mockTailExpiration.toLocaleString(),
+						"[mock expiration date]"
+					)
+			).toMatchInlineSnapshot(`
+			        "Running tail on deploymentId: mock-deployment-id
+			        GET https://example.org/ - Ok @ [mock event timestamp]"
+		      `);
+		});
+
+		it("defaults to logging in json format when the output is not a TTY", async () => {
+			setIsTTY(false);
+
+			const api = mockTailAPIs();
+			await runWrangler(
+				"pages deployment tail mock-deployment-id --project-name mock-project"
+			);
+
+			const event = generateMockRequestEvent();
+			const message = generateMockEventMessage({ event });
+			const serializedMessage = serialize(message);
+
+			api.ws.send(serializedMessage);
+			expect(std.out).toMatch(deserializeToJson(serializedMessage));
+		});
+
+		it("logs console messages and exceptions", async () => {
+			setIsTTY(true);
+			const api = mockTailAPIs();
+			await runWrangler(
+				"pages deployment tail mock-deployment-id --project-name mock-project"
+			);
+
+			const event = generateMockRequestEvent();
+			const message = generateMockEventMessage({
+				event,
+				logs: [
+					{ message: ["some string"], level: "log", timestamp: 1234561 },
+					{
+						message: [{ complex: "object" }],
+						level: "log",
+						timestamp: 1234562,
+					},
+					{ message: [1234], level: "error", timestamp: 1234563 },
+				],
+				exceptions: [
+					{ name: "Error", message: "some error", timestamp: 1234564 },
+					{ name: "Error", message: { complex: "error" }, timestamp: 1234564 },
+				],
+			});
+			const serializedMessage = serialize(message);
+
+			api.ws.send(serializedMessage);
+			expect(
+				std.out
+					.replace(
+						new Date(mockEventTimestamp).toLocaleString(),
+						"[mock event timestamp]"
+					)
+					.replace(
+						mockTailExpiration.toLocaleString(),
+						"[mock expiration date]"
+					)
+			).toMatchInlineSnapshot(`
+			        "Running tail on deploymentId: mock-deployment-id
+			        GET https://example.org/ - Ok @ [mock event timestamp]
+			          (log) some string
+			          (log) { complex: 'object' }
+			          (error) 1234"
+		      `);
+			expect(std.err).toMatchInlineSnapshot(`
+			        "[31mX [41;31m[[41;97mERROR[41;31m][0m [1m  Error: some error[0m
+
+
+			        [31mX [41;31m[[41;97mERROR[41;31m][0m [1m  Error: { complex: 'error' }[0m
+
+			        "
+		      `);
+			expect(std.warn).toMatchInlineSnapshot(`""`);
+		});
+	});
+});
+
+/* helpers */
```

Comment:
some of these are identical to wrangler's `tail` test helpers. i wonder if we could extract and reuse 🤔 . Or could that potentially become a problem in the future?

Reply:
I think for test helpers I'll keep them separated. However, for some of the duplicated controller code, I think we should unify